Cython/Compiler/ModuleNode.py

```diff
@@ -450,10 +450,10 @@ def generate_c_code(self, env, options, result):
         # initialise the macro to reduce the code size of one-time functionality
         code.putln(UtilityCode.load_as_string("SmallCodeConfig", "ModuleSetupCode.c")[0].strip())
 
-        self.generate_module_state_start(env, globalstate['module_state'])
-        self.generate_module_state_defines(env, globalstate['module_state_defines'])
-        self.generate_module_state_clear(env, globalstate['module_state_clear'])
-        self.generate_module_state_traverse(env, globalstate['module_state_traverse'])
+        self.generate_module_state_begin(env, globalstate['module_state'])
+        self.generate_module_state_defines_begin(env, globalstate['module_state_defines'])
+        self.generate_module_state_clear_begin(env, globalstate['module_state_clear'])
+        self.generate_module_state_traverse_begin(env, globalstate['module_state_traverse'])
```

Comment:
I think the function/struct/etc. header parts of the sections can be created in `GlobalState.initialize_main_c_code()`, just not the actual content. Maybe some of these methods can be removed or renamed once they only fill in certain fields or code, or can be merged with their `_end` companion (some of which could donate a few code lines to `GlobalState.close_global_decls()`.

Reply:
Yes I also think this is more complex than necessary in its current form. As I mentioned in one of my other comments somewhere, I would like to come back and just maintain a list of globals in the code object. And then autogenerate the clear and traverse parts from that. (The defines part gets removed naturally, as it shrinks with every change in this branch until it is empty).

But I have to be careful not to extend the scope of these changes more than necessary. I'll come back to this when I have everything in the state struct working.